### ISSUE TITLE
Unable to get IdentityToken to authorize in WebApi

### ISSUE THREAD

**Issue by kkfrosty** [IdentityServer/IdentityServer3.AccessTokenValidation#39]
Simple WebApi setup.  We have a MVC client application which uses Identity Server 3 to authenticate clients.  That portion works fine.

Now we're trying to implement a WebApi project which will use the Token received by clients of our MVC app to authenticate and authorize.  I've dug around and I've looked at the Simplest OAuth2 walk through which seems straight forward.

So I created a console app to test and we use the HttpClient.

Here's a method we build the HttpClient and set the Bearer token.

 public static HttpClient BuildIdentityHttpClient()
        {
            HttpClient _httpClient = new HttpClient();

            //var _client = new OAuth2Client(
            //        new Uri("Https://localhost/IdentityApis/core/connect/token"),
            //        "SLWebApiSvcs_ClientId",
            //        "22ECC1D5-772B-472B-8EBB-2688FBBAEA12");

            //var _token = _client.RequestClientCredentialsAsync("slwebapisvcs").Result;

            var _client = new OAuth2Client(
                //  new Uri("https://localhost/IdSvr/core/connect/token")
               new Uri("Https://localhost/IdentityApis/core/connect/token"),
               "SmartLynxWcfSvcs-ResourceOwner_ClientId",
               "B6E3037C-E8E0-4AEF-BAD0-33A42BA54207"
               );

            var _token = _client.RequestResourceOwnerPasswordAsync("testuser", "P@ssw0rd", "slwebapisvcs").Result;

            _httpClient.SetBearerToken(_token.AccessToken);

            return _httpClient;
        }

After the client has the token, then we make a call like this.

var _result = _client.GetStringAsync("https://localhost/SLWebApi/TruckApi/Get").Result;

On the WebApi side, 

We set the UseIdentityServerBearerTokenAuthentication like so.

  app.UseIdentityServerBearerTokenAuthentication(new IdentityServerBearerTokenAuthenticationOptions
            {
                Authority = "https://localhost/IdentityApis/core",
                RequiredScopes = new[] { "scope1" }
            });

We set the AuthorizeAttribute filter.

Then we use a custom AuthorizeAttribute where we will override the IsAuthorized once we get this figured out.

However, the principal is always Windows principle and there are never claims which were passed over.

We're checking the identity server logs and the WebApi project isn't making calls to Identity server.

When I look at the request object in the customer AuthorizeAttribute class, I can see in the request that authentication is bearer and you see the token.

Why would UseIdentityServerBearerTokenAuthentication not initiate a call to identity server and try to authorize the token and build an ClaimsPrincipal object?

Thanks

**Comment by cbeall** [comment]
Just to help isolate the issue, removing the client console app from the equation, when the WebApi application goes through its startup, it should be calling the discovery endpoint. Is that happening?

**Comment by kkfrosty** [comment]
Thank you for the response Craig.  Everything is hosted local so I don't think a sniff is going to detect this.  I cleared the Identity Server log and just opened a link in a browser.  (Didn't have a token obviously but trying to see if there is anything logged between the server and api.)

In my custom Authorize class, I just return true from the IsAuthorized class so any call passes authorization for testing.  I make the WebApi directly from the browser and I see nothing in the Idenity Server log.

How else would you go about trying to see if the call is being made?  Note I did change the Authority property to point to a bogus URL just so I could run wireshark and see if an attempt was mad to call the made up URL but I saw nothing.  

Having said that if I change the host for the WebApi to IIS Express, then I have verified the Startup is running and all the settings are getting made.

**Comment by cbeall** [comment]
OK... then something is probably not right with settings between web api application and the application running IDS3. You might want to pull the AccessTokenValidation repo down, hook that up to your web api application, and just step through it to troubleshoot. I would have expected an exception to occur during startup, but I'm not exactly sure about that.

Also if it helps, if you don't want to depend on logs, you can add a use statement to middleware before adding IdentityServer and set a breakpoint on the return next().

```
app.Use((context, next) =>
{
	return next();
});
```

**Comment by kkfrosty** [comment]
cbeal, FYI.  I actually spent a lot of time on this, mainly coming up with a way to get WCF to support JWT.  The solution has a WebApi call as well, I've used the same code from it.  Here's a link to the article.  https://kkfrost.wordpress.com/2015/03/05/identity-server-v3-lob-part-2-web-api-wcf-service-integration/
From what I recall, WebApi was fairly simple.

FYI.  our IDS3 server is hosted in IIS and again see the above that it's working with our MVC app & Wcf services using the methods in the blog posts above.

**Comment by kkfrosty** [comment]
Ok, cbeal, I downloaded the AccessTokenValidation project.  I removed the nuget package and added a project reference.  Long story short, this worked.  So I went back to NuGet to reinstall the package and noticed there is a new one released this month.  Namespace is different and I installed that instead of the Thinktecture version from earlier this year.

For whatever reason, everything appears to be working.  I have no idea why and this WebApi project is new and nothing other than a controller with a test action in it.

I wish I could share the technical reason for getting this to work.  I will note again though I think you will see a Start Token request validation if AccessTokenValidation is working.
